### PR TITLE
Addressing #168

### PR DESCRIPTION
**Pull request by mjp41**

### CODE REVIEW

**Comment by mjp41** [comment]
Addresses #168, @SchrodingerZhu does this fix it for you?

**Comment by SchrodingerZhu** [comment]
I confirm the fix.

File: src/ds/defines.h [hunk 1]
```diff
@@ -62,10 +62,15 @@ namespace snmalloc
...
 #    define SNMALLOC_ASSUME(x) \
-      do \
-      { \
-      } while (0)
+      if (!(x)) \
+        __builtin_unreachable();
+#  else
+do
+{
+} while (0)
```
Comment (plietar) [inline src/ds/defines.h:74]: This is missing the #define and backlashes.

Does this path even get used? `__GNUC__ˋ includes both GCC and clang.

Comment (mjp41) [reply src/ds/defines.h:74]: Should have realised, was very confused by clangformat on that one.

Comment (mjp41) [reply src/ds/defines.h:74]: Not used by anything yet.

**Comment by mjp41** [comment]
So we are looking up a value in a table that is guaranteed to be less than `NUM_SMALL_SIZECLASSES` I guess it cannot determine that this is the case.  It would be a fairly complex static analysis to get this right.  I seem to remember @davidchisnall said this analysis is conservative, it can loose precision and will report errors based on that.  The `SNMALLOC_ASSUME` stuff was helping but I guess that is not the case with more recent GCC.

Thought, it it strange that your PR caused the issue.  I wonder if the code pointer dispatch in `function_ref` is causing a lose of precision in the analysis, which is strange as the table is `constexpr`.

**Comment by plietar** [comment]
I did a bit more digging, and managed to get the following minimal reproduction:
```c++
#include <cstddef>

bool maybe1();
bool maybe2();

class C
{
  void* data[42];
  int extra;

public:
  void* foo(size_t index)
  {
    if (index < 42)
    {
      return nullptr;
    }
    if (maybe1())
    {
      return bar(index);
    }
    return nullptr;
  }

  void* bar(size_t index)
  {
    if (maybe2()) {
      return data[index];
    }
    return nullptr;
  }
};

int call_foo(C* obj, int index)
{
  obj->foo(index);
  return 0;
}
```

From what I was able to find online, the GCC warning is defined as "assuming the array access is reached, it is always OOB", rather than "this array access is reachable and is out-of-bounds".

In the code above, I think GCC inlines foo and bar into call_foo *before* doing the OOB analysis, and figures out that if the `data[index]` line is reached (ie. maybe1 and maybe2 are true) then it must be OOB. It makes sense in the inlined version, but in the original version `bar` could have been called directly, with a valid index. That's a pretty weird behaviour and probably a bug in GCC.

> Though, it it strange that your PR caused the issue. I wonder if the code pointer dispatch in function_ref is causing a lose of precision in the analysis, which is strange as the table is constexpr.

I think it's the other way around, `function_ref` gets inlined better than `std::function`, so GCC finds the OOB condition (from the large alloc code path recursing into the small alloc one) whereas it didn't before.

**Comment by mjp41** [comment]
So do we have an issue on that path?

**Comment by mjp41** [comment]
So I had a bit of a play based on your observation.  If we call back with `small_first_alloc` calling back in to `small_alloc_inner`, and `medium_alloc` into `medium_alloc` and `large_alloc` into `large_alloc` then the warning goes away.

In doing this, I also released the `zero_mem` flag is not correctly set for the call-back. So I'll add a test for this, and adjust the code, and reenable the warning.

**Pull request merged**